### ISSUE TITLE
Option to automatically suspend tasks if estimated time > deadline + X hours

### ISSUE THREAD

**Issue by cristipurdel** [BOINC/boinc#2957]
**Describe the problem**
As per title

**Describe the solution you'd like**
I would like to have an option to check every e.g. 15 minutes, if estimated time > deadline + X hours, abort the task automatically.
So that I do not do this manually, nor do I complete the task but not receive any credit.

**Additional context**
Option should be disabled by default.
I have a low Nvidia card running GPUGRID tasks, and sometimes estimated time > deadline.
Similar to https://github.com/BOINC/boinc/issues/2115

**Comment by Ferroin** [comment]
This would also help BOINC systems that have been offline for an extended period of time get back to actual useful crunching of work in a reasonable amount of time.

**Comment by ethanteng** [comment]
I was thinking of picking this up as it'd be a nice follow-on to #2115 that I just implemented. But, I wanted to propose an alternative solution and get some feedback. Instead of automatically aborting at deadline + X hours (configurable), what if BOINC automatically _suspended_ a task *as soon as* the estimated_ time > deadline? Then, it'd be up to the user to either abort or resume the task. To me, that seems to be a more intuitive behavior. Thoughts?

**Comment by AenBleidd** [comment]
I ask you to wait for @davidpanderson to implement #2993. Otherwise either you or David will do an extra work.

**Comment by davidpanderson** [comment]
There's value (and credit) in finishing jobs even if they miss their deadline.
Automatically suspending jobs could lead to unbounded suspended jobs.

**Comment by ethanteng** [comment]
@davidpanderson What about @cristipurdel's original suggestion of automatically _aborting_ a job if the estimated completion date is X hours past the deadline, where X is user configurable? Would aborting be preferable to suspending?

Or, is your point that users _shouldn't_ do either (abort or suspend) and instead, let the job continue to run and finish because there is still value (and credit) in doing so?

**Comment by AenBleidd** [comment]
@davidpanderson, according to [this](https://boinc.berkeley.edu/trac/wiki/ProjectOptions) documentation, there are two options:
- `report_grace_period`
- `grace_period_hours`
that have next description:
`A "grace period" (in seconds or hours respectively) for task reporting. A task is considered time-out (and a new replica generated) if it is not reported by client_deadline + x.`
As far as I know, client knows about client_deadline only. So maybe it worth to send to the client information about grace period and cancel task that failed to finish in time including this grace period? Because afterwards it's just a wasted CPU time.

**Comment by RichardHaselgrove** [comment]
I think we ought to consider long and deep before we interfere with the multiple, sometimes quite subtle, safeguards already built into BOINC.

There are two audiences we need to consider: the active, involved, participant, who thinks they know how BOINC works and wants to get the best out of it. And the detached, 'set and forget', participant who loads the software and leaves it alone. There are gradations between these extremes, and people sometimes move more towards one extreme or the other as life circumstances change.

For active, involved, participants: why would they get into deadline trouble in the first place? If we do anything, we should ask that question at an earlier stage. Why would someone attach a low-end video card to GPUGrid? The have a 5-day deadline, but actively encourage users to return work within 24 hours - it's the nature of their science.

For set and forget users, such as Science United: David is actually wrong - there will be no unbounded backlog of job, because work fetch is inhibited for any project with a suspended task. Hosts will simply drop out of productive participation in any affected project.

My initial starting point for discussion might be: we should inhibit work fetch for any project/resource pairing that has a task in deadline trouble (running in EDF). We may even do that already - I'm getting rusty on the subtleties myself.

**Comment by AenBleidd** [comment]
@RichardHaselgrove, implement such option disabled by default is safe enough, as for me

**Comment by RichardHaselgrove** [comment]
@AenBleidd - I'd been thinking along similar lines myself. The point where it becomes wasteful to continue with a task is when a workunit is finalised and assimilated. The replacement when a task goes beyond deadline goes through many delays:

* Any grace period set by the project
* waiting at end of 'ready to send' server queue
* allocation to a suitable host (consider locality scheduling)
* waiting in local host cache before running
* and finally, actually computing on the replacement host

The server has a concept of 'result is no longer usable', shown in acknowledgement of completed tasks. Maybe that status should be used to trigger an unconditional abort?

---

Re the 'option disabled by default' - that's why I made the point about people's involvement varying over time. What if you set the option, but then go away on holiday? Or into hospital?

**Comment by AenBleidd** [comment]
> Maybe that status should be used to trigger an unconditional abort?

Yeah, this could be an option, you're right.

> Re the 'option disabled by default' - that's why I made the point about people's involvement varying over time. What if you set the option, but then go away on holiday? Or into hospital?

You mean you set this option 'on'? I believe BOINC will just cancel spoiled tasks. I think this option is of that kind of options that are just set and never changed for years.

**Comment by RichardHaselgrove** [comment]
Which is why we need an ultimate, final, disposal route - not just indefinite suspension.

**Comment by ethanteng** [comment]
I like the idea of triggering an unconditional abort when the result is no longer usable. Can someone point me to where this is determined in the server code? I can then look into how we could possibly leverage that in the client.

**Comment by RichardHaselgrove** [comment]
The client already handles automatic aborts, both unconditional and 'abort if not started'. The only work that would be needed is on the server. I suggest it's tied in with 'resend lost results': that's an optional feature which works well if the server is powerful enough to handle the extra workload, but can cripple a very heavily loaded database. A database expert will have to advise whether the additional query (status of workunit, after checking existence of allocated task in sched request list) can be accommodated efficiently.

**Comment by RichardHaselgrove** [comment]
Read from https://github.com/BOINC/boinc/blob/master/sched/handle_request.cpp#L674

**Comment by ethanteng** [comment]
What if the user configurable option is something like: "**Exclude jobs if it is at risk of missing its deadline (Y/N)**"?

If the user sets this to "true", then the following logic would apply:

**In cpu_sched->can_schedule(): if result->rr_sim_misses_deadline == true, then return false.**

rr_sim_misses_deadline is set by rr_sim->simulate() and is set to true if the result is expected to miss the deadline.

rr_sim->simulate() is called within rr_sim->rr_simulation(), which is itself called within cpu_sched->make_run_list() and upwards, from cpu_sched->schedule_cpus(). This is called by client_state->poll_slow_events(), which is ultimately and intermittently called within main().

So tl;dr: whenever main() determines the list of tasks to run - and the order to run them - any tasks that are at risk of slipping past its deadline is excluded, if the user has configured it to be so.

Thoughts on this approach? I'd love some feedback before I create a PR.

**Comment by davidpanderson** [comment]
There are much more important things to work on, e.g. #2273

**Comment by ethanteng** [comment]
@davidpanderson Undoubtedly there are more important things. However, I'm new to this project and am looking for work to (a) help familiarize myself with the codebase, and (b) make an immediate impact / add value. This seemed like a good follow-up to the work I did for #3403, hence my interest in picking it up.

**Comment by CharlieFenton** [comment]
Thank you for wanting to contribute to BOINC, but I wonder if there are other tasks that might be a better choice for your first effort, especially since the scheduler and the way the client and server interact for scheduling are among the most complex parts of BOINC. I believe there are long wish lists of changes that would be very helpful.

**Comment by AenBleidd** [comment]
Closed as a duplicate of #2366
